### PR TITLE
WT-10757 Add an I/O trace visualization tool 

### PR DESCRIPTION
**Pull request by pmacko86**

### CODE REVIEW

**Comment by ddanderson** [comment]
@pmacko86, I'm in the middle of reviewing, but I don't have experience with Gtk.  Might be good to ask on storage engines to see if there's anyone that knows some gtk to review that part.

File: tools/io-trace-explorer/io_trace.cc [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+        int count = 0;
+        char *parts[16];
+        char *next(buf), *p;
+        while ((p = strsep(&next, " \t\n\r")) != NULL) {
+            if (*p == '\0')
+                continue;
+            parts[count++] = p;
+            if (count >= 16)
+                break;
```
Comment (ddanderson) [inline tools/io-trace-explorer/io_trace.cc:185]: Maybe use  WT_ELEMENTS(parts) or use the same sizeof trick to avoid having 16 in two places.  (Or use const size value).

File: tools/io-trace-explorer/io_trace.cc [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+     * [1678914463:713343][29429:0x7f532741cc40], file:WiredTiger.wt, WT_SESSION.create:
+     *   [WT_VERB_WRITE][DEBUG_2]: write: /data/wt/./WiredTigerLog.0000000001, fd=7,
+     * offset=1920, len=2048 [1678914463:715599][29429:0x7f53157fa640]:
+     * [WT_VERB_WRITE][DEBUG_2]: write: /data/wt/./WiredTigerLog.0000000001, fd=7, offset=3968,
+     * len=131328
+     */
+
+    char buf[256];
+    double first_timestamp = 0;
+    unsigned lines = 0;
```
Comment (ddanderson) [inline tools/io-trace-explorer/io_trace.cc:314]: I think we prefer `unsigned int` in our codebase (I assume that was the intention).

File: tools/io-trace-explorer/io_trace.cc [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+     *   9) Block length
+     *  10) Duration in (), in microseconds (enabled via the "-t" command-line argument)
+     *  11) Process name and potentally other process-relevant information in []
+     *
+     * Note that actions other than C and D can carry different extra information on fields 7
+     * and up.
+     */
+
+    char buf[256];
+    unsigned lines = 0;
```
Comment (ddanderson) [inline tools/io-trace-explorer/io_trace.cc:169]: Prefer `unsigned int`

File: tools/io-trace-explorer/io_trace.cc [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+        if (write &&
+          (strcmp(parts[index + 1], "write") != 0 && strcmp(parts[index + 1], "write-mmap") != 0))
+            continue;
+
+        /* Parse the line, starting with the timestamps. */
+
+        char *e;
+        io_trace_operation item;
+        bzero(&item, sizeof(item));
```
Comment (ddanderson) [inline tools/io-trace-explorer/io_trace.cc:370]: Maybe it's better having a constructor for io_trace_operation do the bzero?

**Review by ddanderson — APPROVED** [review]
Looks good to me, modulo a couple small items.  I didn't look very closely at the GTK stuff, hoping to leave that to someone more capable.  There is the matter of formatting a data partition that I think we should be very careful with.

**Comment by luke-pearson** [comment]
@pmacko86 WiredTiger recently agreed upon a C++ style which can be found on our wiki. Could you please bring the code in-line with that style?

File: tools/io-trace-explorer/io_trace.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    /*
+     * wrap_timestamp --
+     *     Wrap a timestamp into an instance of io_trace_operation, which is useful for filtering a
+     * collection of traces.
+     */
+    static io_trace_operation
+    wrap_timestamp(double t)
+    {
+        io_trace_operation r;
+        bzero(&r, sizeof(r));
```
Comment (luke-pearson) [inline tools/io-trace-explorer/io_trace.h:54]: I was unfamiliar with `bzero`, upon googling it appears to set the memory of the given struct to all zeros. I believe it is a deprecated and POSIX only. In C++ the standard way to zero a struct is: `r = {}`.

File: tools/io-trace-explorer/io_trace.cc [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+io_trace_collection::add_data_point(
+  const std::string &device_or_file, const io_trace_kind kind, const io_trace_operation &item)
+{
+    std::string name = device_or_file;
+
+    /* Postprocess the data point. */
+
+    if (kind == io_trace_kind::WIRED_TIGER) {
+        const char *s = name.c_str();
+        if (strstr(s, "Log.00000") || strstr(s, "log.00000")) {
```
Comment (luke-pearson) [inline tools/io-trace-explorer/io_trace.cc:59]: Single line if statements shouldn't be bracketed.

Comment (luke-pearson) [inline tools/io-trace-explorer/io_trace.cc:59]: We prefer to not mix C style string methods in with C++ code. 
- strstr -> name.find()
- strlen -> name.length()
- strcmp -> name.compare() or == if not requiring a substring match

File: tools/io-trace-explorer/io_trace.cc [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+              std::upper_bound(t->_operations.begin(), t->_operations.end(), item), item);
+    }
+}
+
+/*
+ * io_trace_collection::load_from_file --
+ *     Load a collection from file.
+ */
+void
+io_trace_collection::load_from_file(const char *file)
```
Comment (luke-pearson) [inline tools/io-trace-explorer/io_trace.cc:107]: It looks to me like all callers to this have a std::string, any reason `char *` is preferable here?

File: tools/io-trace-explorer/io_trace.cc [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+{
+    FILE *f = fopen(file, "r");
```
Comment (luke-pearson) [inline tools/io-trace-explorer/io_trace.cc:109]: In c++ `ofstream`, `fstream` and `ifsteam` are the standard file read / write / open mechanisms. I am not too fussed about using `fopen` but if you get time I suggest replacing this too.

Comment (pmacko86) [reply tools/io-trace-explorer/io_trace.cc]: I rewrote loading using the C++ file and string parsing primitives (e.g., `ifstream`), but the loading then slowed down by more than 2x, so I reverted those changes. But the code was definitely neater.

If there is no objection, let's just keep the original C loading and parsing for the sake of performance—the loading time is already ~50 seconds for a 1.5 hour trace.

**Review by ddanderson — APPROVED** [review]
Looks good - I noted a couple nits.
Thanks for resolving my previous comments.

File: tools/io-trace-explorer/io_trace.cc (hunk 8 above)
Comment (luke-pearson) [reply tools/io-trace-explorer/io_trace.cc]: That sounds reasonable to me, I'll try and get to reviewing the rest this afternoon / monday.

File: tools/io-trace-explorer/README [hunk 9]
```diff
@@ -0,0 +1,55 @@
+ I/O Trace Explorer
+====================
+
+I/O Trace Explorer is a GUI tool for visualizing I/O traces from WiredTiger, at both the file level
+and at the device level. It is useful for helping us to better understand and optimize our I/O
```
Comment (luke-pearson) [inline tools/io-trace-explorer/README:5]: I would rephrase this to say: "It helps us better understand and optimize our I/O patterns".

File: tools/io-trace-explorer/README [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+ Capturing traces
+------------------
+
+To capture traces at the file level: Enable DEBUG_2 level logging for WT_VERB_READ and
+WT_VERB_WRITE and capture logs to a file, e.g., by piping the stdout to a file. You can enable the
+read and write logging by running the following before the workload:
+
+    export WIREDTIGER_CONFIG="verbose=[read:2,write:2]"
+
+To capture traces at the device level: Use blktrace to capture the block trace at the device level,
```
Comment (luke-pearson) [inline tools/io-trace-explorer/README:30]: You can probably delete "To capture traces at the device level:"

File: tools/io-trace-explorer/io_trace.cpp [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+        t = new io_trace(*this, name.c_str());
+        _traces[name] = t;
+    } else
+        t = ti->second;
+
+    /* Insert the item into the correct place to keep the list sorted. */
+
+    if (t->_operations.empty())
+        t->_operations.push_back(item);
+
```
Comment (luke-pearson) [inline tools/io-trace-explorer/io_trace.cpp:89]: I would remove this whitespace line, it offsets the else { weirdly.
